src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -360,13 +360,65 @@ private BoundStatement BindThrow(ThrowStatementSyntax node, DiagnosticBag diagno
             return new BoundThrowStatement(node, boundExpr, hasErrors);
         }
 
-        private static BoundStatement BindEmpty(EmptyStatementSyntax node)
+        private static bool IsEmbeddedStatementParent(SyntaxNode node)
         {
+            if (node != null)
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.FixedStatement:
+                    case SyntaxKind.DoStatement:
+                    case SyntaxKind.IfStatement:
+                    case SyntaxKind.ElseClause:
+                    case SyntaxKind.LockStatement:
+                    case SyntaxKind.UsingStatement:
+                    case SyntaxKind.ForStatement:
+                    case SyntaxKind.ForEachStatement:
+                    case SyntaxKind.WhileStatement:
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static BoundStatement BindEmpty(EmptyStatementSyntax node, DiagnosticBag diagnostics)
+        {
+            if (IsEmbeddedStatementParent(node.Parent))
+            {
+                switch (node.Parent.Kind())
+                {
+                    default:
+                        // For non-loop constructs, always warn.  This is for code like:
+                        // "if (x) ;" which is almost certainly a bug.
+                        diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation());
+                        break;
+
+                    case SyntaxKind.ForStatement:
+                    case SyntaxKind.ForEachStatement:
+                    case SyntaxKind.WhileStatement:
+                        if (node.SemicolonToken.GetNextToken().Kind() == SyntaxKind.OpenBraceToken)
+                        {
+                            // For loop constructs, only warn if we see a block following the statement.
+                            // That indicates code like:  "while (x) ; { }"
+                            // which is most likely a bug.
+                            diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation());
+                        }
+                        break;
+                }
+            }
+
             return new BoundNoOpStatement(node, NoOpStatementFlavor.Default);
         }
 
         private BoundLabeledStatement BindLabeled(LabeledStatementSyntax node, DiagnosticBag diagnostics)
         {
+            if (IsEmbeddedStatementParent(node.Parent))
```

Comment:
>if (IsEmbeddedStatementParent(node.Parent)) [](start = 12, length = 43)

It feels like it would be better to intercept this statement in BindPossibleEmbeddedStatement function instead.